openapi_python_client/parser/properties/union.py

```diff
@@ -189,3 +199,105 @@ def validate_location(self, location: oai.ParameterLocation) -> ParseError | Non
+            )
+        name = top_level_model.name
+        if name.startswith("/components/schemas/"):
+            name = get_reference_simple_name(name)
+        model_types_by_name[name] = top_level_model
+
+    # The discriminator can specify an explicit mapping of values to types, but it doesn't
+    # have to; the default behavior is that the value for each type is simply its name.
+    mapping: dict[str, PropertyProtocol] = model_types_by_name.copy()
+    if data.mapping:
+        for discriminator_value, model_ref in data.mapping.items():
+            ref_path = parse_reference_path(
+                model_ref if model_ref.startswith("#/components/schemas/") else f"#/components/schemas/{model_ref}"
+            )
+            if isinstance(ref_path, ParseError) or ref_path not in schemas.classes_by_reference:
+                return PropertyError(detail=f'Invalid reference "{model_ref}" in discriminator mapping')
+            name = get_reference_simple_name(ref_path)
+            if not (lookup_model := model_types_by_name.get(name)):
+                return PropertyError(
+                    detail=f'Discriminator mapping referred to "{model_ref}" which is not one of the schema variants',
+                )
+            for original_value in (name for name, m in model_types_by_name.items() if m == lookup_model):
+                mapping.pop(original_value)
+            mapping[discriminator_value] = lookup_model
+    else:
+        mapping = model_types_by_name
+
+    return DiscriminatorDefinition(property_name=data.propertyName, value_to_model_map=mapping)
+
+
+def _validate_discriminators(
+    discriminators: list[DiscriminatorDefinition],
+) -> PropertyError | None:
+    from .model_property import ModelProperty
+
+    prop_names_values_classes = [
+        (discriminator.property_name, key, cast(ModelProperty, model).class_info.name)
+        for discriminator in discriminators
+        for key, model in discriminator.value_to_model_map.items()
+    ]
+    for p, v in {(p, v) for p, v, _ in prop_names_values_classes}:
+        if len({c for p1, v1, c in prop_names_values_classes if (p1, v1) == (p, v)}) > 1:
+            return PropertyError(f'Discriminator property "{p}" had more than one schema for value "{v}"')
+    return None
+
+    # TODO: We should also validate that property_name refers to a property that 1. exists,
+    # 2. is required, 3. is a string (in all of these models). However, currently we can't
+    # do that because, at the time this function is called, the ModelProperties within the
+    # union haven't yet been post-processed and so we don't have full information about
+    # their properties. To fix this, we may need to generalize the post-processing phase so
+    # that any Property type, not just ModelProperty, can say it needs post-processing; then
+    # we can defer _validate_discriminators till that phase.
```

Comment:
"Post-processed" here refers to the two-pass behavior that the generator uses to deal with forward references. In the initial parsing phase, the `ModelProperty` instances created for object schemas do not contain any actual property information— because parsing the properties might require resolving a `$ref` to a type that hasn't been parsed yet. So then after all the types have been seen, it does a second pass where it finishes parsing the properties of every `ModelProperty`.